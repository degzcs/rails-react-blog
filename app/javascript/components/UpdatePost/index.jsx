import React , { useState, useEffect } from 'react';
import { useHistory, useParams } from 'react-router-dom'
import PostForm from '../PostForm/index.jsx'

const UpdatePost  = ({handleError, clearError}) => {
  const [post, setPost] = useState({
    title: '',
    description: '',
    content: '',
    image: null
  })
  const history = useHistory()
  const { slug } = useParams()

  useEffect(() => {
    const getPost = async () => {
      const postFromServer = await fetchPost(slug)
      setPost({...postFromServer, image: null})
    }

    getPost()
  }, [])

  // Fetch Post
  const fetchPost = async (slug) => {
    const res = await fetch(`http://127.0.1:3000/api/v1/posts/show/${slug}`)
    const data = await res.json()

    return data
  }

  // Update
  const update = async () => {
    const token = document.querySelector('meta[name="csrf-token"]').content;
    const formData = new FormData()
    formData.append('title', post.title)
    formData.append('content', post.content)
    formData.append('description', post.description)
    formData.append('slug', post.slug)
    if(post.image) formData.append('image', post.image)

    const res = await fetch('http://127.0.0.1:3000/api/v1/posts/update', {
      method: 'PUT',
      headers: {
                'X-CSRF-Token': token
              },
      body: formData
    }).catch( e => {
      handleError(e)
    })

    if(!res) return;
    clearError()
    const data = await res.json()
    history.push(`/blog/${data.slug}`)
  }

  // Events
  const onSubmit = (e) => {
    e.preventDefault()
    update()
  }

  const onChange = (e) => {
    setPost({...post, [e.target.name]: e.target.value });
  }

  const handleFileChange = (e) => {
    setPost({...post, [e.target.name]: e.target.files[0]});
  }

  return(
    <>
    <PostForm
      post={post}
      onChange={onChange}
      onSubmit={onSubmit}
      handleFileChange={handleFileChange}
      newPost={false}
    />
    </>
  )
}

export default UpdatePost
