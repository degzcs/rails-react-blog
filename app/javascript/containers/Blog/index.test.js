import React from "react";
import renderer from "react-test-renderer";
import { StaticRouter } from "react-router";

import UpdatePost from "./index.jsx";

test("checks rendering of UpdatePost component", () => {
  const component = renderer.create(
    <StaticRouter location="someLocation">
      <UpdatePost />
    </StaticRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
