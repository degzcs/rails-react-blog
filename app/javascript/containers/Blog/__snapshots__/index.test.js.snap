// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks rendering of UpdatePost component 1`] = `
<div
  className="container"
>
  <div
    className="py-5"
  >
    <main
      className="container-fluid"
    >
      <div
        className="row mt-5"
      >
        <ul
          className="nav justify-content-end"
        >
          <li
            className="nav-item"
          >
            <a
              className="btn btn-lg bg-light"
              href="/blog/new-post/"
              onClick={[Function]}
            >
              New Post
            </a>
          </li>
        </ul>
      </div>
      <div
        className="row posts mb-5"
      >
        <div>
          <p>
            Loading...
          </p>
        </div>
      </div>
    </main>
  </div>
</div>
`;
